### PR TITLE
add blog post how to migrate databases

### PR DESCRIPTION
**Pull request by timogoebel**
... this is still WIP as I need to check the commands. Already putting this out here in case anybody wans to check my English. :-)

cc: @lzap, @GregSutcliffe as you asked for this iirc.

### CODE REVIEW

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+<!--more-->
+While this blog post describes the required steps to migrate from MySQL to Postgresql, it can easily be adapted to migrate from Postgresql to MySQL.
+
+## Be careful
+
+First of all, a word of caution: This is not for the faint-hearted. And even fearless sysadmins should take backups before starting with this task. There is no guarantee, this will work.
+
+## Motivation
+
+Postgresql and MySQL are both battle-tested open source database management systems. They have specific advantages and disadvantages. The most important reason to prefer one over the other is the available know-how in your organization. If you have a dedicated MySQL team that can provide a database for you, then you should by all means prefer.
```
Comment (ekohl) [inline _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md]: The prefer at the end feel like it misses ```that```.

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+First of all, you need to install support for either Postgresql or MySQL to Foreman. One of them will be missing. `yum install -y foreman-postgresql` adds support for Postgresql, `yum install -y foreman-mysql2` does the same for MySQL.
+
+Next, stop all your foreman instances and make sure, they stay stopped for good. On RedHat systems, you need to stop the `httpd` and (if applicable) `foreman-tasks` service.
+
+You can issue the MySQL command `SHOW FULL PROCESSLIST;` to show you all active connections to your MySQL database. Make sure, there are none.
+
+If you are using the foreman-tasks plugin, you need to patch the script we are using to migrate the data. If you are not sure, wether you use this plugin, check if you have an entry `Tasks` in Foreman's Monitor menu. If you are still not sure, it is safe to apply the patch in any case.
+
+```sh
+wget -O /tmp/convert-tasks-patch.diff https://gist.github.com/...
```
Comment (ekohl) [inline _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md]: The URL is missing something. I'd also add a note if this is only needed in 1.15 or similar.

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+yum install patch -y
+cd /usr/share/foreman
+
+patch -p1 < /tmp/convert-tasks-patch.diff
+```
+
+## Migrating
+
+To migrate the data from one database to the other, we will use a rake task. You might already be familiar to this tasks from upgrading Foreman to a newer version.
```
Comment (ekohl) [inline _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md]: s/this tasks/this task/

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+Next, we can finally migrate all the data. This will take a long time depending on the size of your database. It is highly recommended to run this in a screen or tmux session.
+
+```sh
+forman-rake db:convert:prod2dev
+```
+
+## Wrapping up
+
+Postgresql uses sequences in order to store the next free primary key id for a table. These
```
Comment (ekohl) [inline _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md]: This sentence is incomplete.

**Review by ekohl — COMMENTED** [review]
Maybe a note about stopping anything that might start services again (puppet, mon, ..., crons) would be good.

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md (hunk 2 above)
Comment (lzap) [reply _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md]: Yeah, also send patch. I guess it is already pending review somewhere :-)

**Comment by lzap** [comment]
Great, thanks so much.

With your rich production experience, I would *love* to see some more in-depth summary of performance, some numbers. Does it perform worse, better, first experience? Is there something you are missing now (e.g. replication). These are all great items to have here.

As a redhatter, I only have experience with PostgreSQL Foreman setups...

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
+---
+layout: post
+title: Migrating Foreman from MySQL to Postgresql
+date: 2017-10-31 11:55:05
+author: Timo Goebel
+tags:
+- foreman
+---
+
+While Foreman supports both MySQL and Postgresql as a database management system, Katello requires you to use Postgresql. If you want to add the Katello plugin on an existing Foreman setup with MySQL, you need to migrate your database. But there are other reasons why you might want to switch your database management system. This blog post describes what steps are required to migrate your data.
```
Comment (sean797) [inline _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md]: s/Katello requires you to use Postgresql/Katello requires Postgresql/ ?

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+Foreman internally uses a development framework called Rails. In Rails, you can have a specific configuration for e.g. a development and a production environment. Our prod2dev rake task assumes, that the source database is configured as the production database and the new target database is configured as the development database. All data in the development database will be overwritten.
+
+You can configure the new database as Foreman's new development database by adding a new connection in `/etc/foreman/database.yml`.
+
+It may look like this:
+
+```yaml
+# The new target database
+development:
+  adapter: sqlite3
```
Comment (sean797) [inline _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md]: Should these examples use pgsql and mysql ?

**Review by sean797 — COMMENTED** [review]
Nice, 2 small comments :-)

I noticed you haven't mentioned the installer, is that intentional?

**Comment by timogoebel** [comment]
@ekohl, @lzap, @sean797: Thanks for your awesome feedback.

I fixed your comments, added the link to the required patch and fixed the examples. Also added some sections about what you suggested.
I forgot, that the user that does the migration needs superuser privileges. Added a section about that.

@lzap: Unfortunately, I don't have any interesting numbers to share. But I tried to describe my first experience. :-(

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+Postgresql and MySQL are both battle-tested open source database management systems. They have specific advantages and disadvantages. The most important reason to prefer one over the other is the available know-how in your organization. If you have a dedicated MySQL team that can provide a database for you, then you should by all means prefer that.
+If you do not have anybody who can provide you with a database on either MySQL or Postgresql, I'd tend to the latter as Postgresql seems a little more robust.
+
+## Prerequisites
+
+This blog post assumes that you already have the source and target databases set up. Make sure to setup the Postgresql database with UTF-8 enabled.
+
+First of all, you need to install support for either Postgresql or MySQL to Foreman. One of them will be missing. `yum install -y foreman-postgresql` adds support for Postgresql, `yum install -y foreman-mysql2` does the same for MySQL.
+
+Next, stop all your foreman instances and make sure, they stay stopped for good. On RedHat systems, you need to stop the `httpd` and (if applicable) `foreman-tasks` service. Make sure that neither Puppet or a monitoring system automatically starts these services again.
```
Comment (ekohl) [inline _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md]: I think this should be ```... neither Puppet nor a ...```

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+You can issue the MySQL command `SHOW FULL PROCESSLIST;` to show you all active connections to your MySQL database. Make sure, there are none.
+
+If you are using the foreman-tasks plugin, you need to patch the script we are using to migrate the data. If you are not sure, wether you use this plugin, check if you have an entry `Tasks` in Foreman's Monitor menu. If you are still not sure, it is safe to apply the patch in any case.
+
+```sh
+wget -O /tmp/convert-tasks-patch.diff https://gist.githubusercontent.com/timogoebel/cd29e5206678ca0e0d78222af70583ec/raw/bbe998624098c1c3e656fe2582d33b0d9f9285aa/0001-convert-skip-foreman-tasks-tables.patch
```
Comment (ekohl) [inline _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:37]: Is this patch upstream included? Should it be noted that it's only for 1.15?

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+Postgresql uses sequences in order to store the next free primary key id for a table. These sequences do not get updated by the rake tasks and have to be adjusted manually. On MySQL, this is not needed as the concept of sequences is not known there.
+
+The Postgresql Wiki has a [script to repair the sequences](https://wiki.postgresql.org/wiki/Fixing_Sequences). Copy the script and save it to a file, e.g. `sequences.sql`. Then run it like this:
+
+```sh
+su - postgres
+vim sequences.sql
+# Paste the script and save the file
+FIXME
```
Comment (ekohl) [inline _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md]: I think there's something missing here :)

**Review by ekohl — CHANGES_REQUESTED** [review]
I'd expect something to create the foreman database or at least the user. Should this be documented too?

File: _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md (hunk 8 above)
Comment (timogoebel) [reply _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:37]: It's not included upstream. We should provide a way for plugins to specify tables that should be skipped. But that's not trivial to do.
dynflow defines tables that are outside of active record's control. As dynflow is now part of core, I believe we could skip all dynflow tables there. That would be a simple change.

**Comment by timogoebel** [comment]
> I'd expect something to create the foreman database or at least the user. Should this be documented too?

@ekohl: I used puppet for that. But users might use something else, e.g. plain SQL. There is already a ton of information on how this is done on the internet. So I thought it was safe to assume, that this is already done.

Fixed your other comments. Thanks.

**Comment by ekohl** [comment]
> @ekohl: I used puppet for that. But users might use something else, e.g. plain SQL. There is already a ton of information on how this is done on the internet. So I thought it was safe to assume, that this is already done.

Since nothing else mentions puppet I'd at least add a note about the installer or ```sudo -u postgres createuser foreman && sudo -u postgres createdb foreman -O foreman -E UTF-8 -l en_US.UTF-8```

**Review by lzap — APPROVED** [review]
Thanks.

**Review by dLobatog — APPROVED** [review]
Since this is already approved & after reading it it looks great, I'm going to go with merging. Thanks @timogoebel & everyone else involved! Time to promote this :)

**Pull request merged**